odoo/modules/registry.py

```diff
@@ -309,19 +309,26 @@ def post_constraint(self, func, *args, **kwargs):
         """ Call the given function, and delay it if it fails during an upgrade. """
         try:
             func(*args, **kwargs)
+            if (func, args, kwargs) in self._constraint_queue:
```

Comment:
Could be:
```python
if (func, args, kwargs) not in self._constraint_queue:
    func(*args, **kwargs)
```
this not only avoids a potentially expensive removal from the queue, but also prevents having multiple entries of the same constraint in the queue (which would trigger the same error as explained in the comment).

The side-effect would be that the constraint would always be applied at the very end of the registry cycle even if subsequent attempts by inheriting modules would succeed

Reply:
> The side-effect would be that the constraint would always be applied at the very end of the registry cycle even if subsequent attempts by inheriting modules would succeed

I don't think that's much of an issue, after all technically the constraints could have been applied long before the end of the registry loading. We have no idea when, but we *could* attempt to apply deferred constraints after the install / update / loading of each module. I think the simpler code and lower amount of explanation wins.